src/main/java/org/apache/hadoop/fs/PrestoFileSystemCache.java

```diff
@@ -13,61 +13,82 @@
 
+import static com.google.common.base.Preconditions.checkNotNull;
+import static java.util.Objects.requireNonNull;
 import static org.apache.hadoop.fs.FileSystem.getFileSystemClass;
+import static org.apache.hadoop.security.UserGroupInformationShim.getSubject;
 
 public final class PrestoFileSystemCache
         extends FileSystem.Cache
 {
     private static final int CACHE_LEAK_PROTECTION_THRESHOLD = Integer.parseInt(System.getProperty("hadoopFileSystemCacheLeakProtectionThreshold", "100000"));
 
     private final AtomicLong unique = new AtomicLong();
-    private final Map<Key, FileSystem> map;
-
-    @SuppressWarnings("unchecked")
-    public PrestoFileSystemCache()
-    {
-        map = getPrivateMap(this);
-    }
+    private final Map<FileSystemKey, FileSystemHolder> map = new HashMap<>();
 
     @Override
     FileSystem get(URI uri, Configuration conf)
             throws IOException
     {
-        return getInternal(uri, conf, new Key(uri, conf));
+        return getInternal(uri, conf, 0);
     }
 
     @Override
     FileSystem getUnique(URI uri, Configuration conf)
             throws IOException
     {
-        return getInternal(uri, conf, new Key(uri, conf, unique.incrementAndGet()));
+        return getInternal(uri, conf, unique.getAndIncrement());
     }
 
-    private synchronized FileSystem getInternal(URI uri, Configuration conf, Key key)
+    private synchronized FileSystem getInternal(URI uri, Configuration conf, long unique)
             throws IOException
     {
-        FileSystem fs = map.get(key);
-        if (fs != null) {
-            return fs;
+        UserGroupInformation userGroupInformation = UserGroupInformation.getCurrentUser();
+        FileSystemKey key = createFileSystemKey(uri, userGroupInformation, unique);
+        Set<?> privateCredentials = getPrivateCredentials(userGroupInformation);
+
+        FileSystemHolder fileSystemHolder = map.get(key);
+        if (fileSystemHolder == null) {
+            FileSystem fileSystem = createFileSystem(uri, conf);
+            fileSystemHolder = new FileSystemHolder(fileSystem, privateCredentials);
+
+            if (map.size() > CACHE_LEAK_PROTECTION_THRESHOLD) {
+                throw new IOException("Hadoop file system leak protection threshold has been reached. Threshold can be changed by setting " +
+                        "'-DhadoopFileSystemCacheLeakProtectionThreshold=<value>' jvm property. Current limit is: " + CACHE_LEAK_PROTECTION_THRESHOLD);
+            }
+
+            map.put(key, fileSystemHolder);
         }
 
-        if (map.size() > CACHE_LEAK_PROTECTION_THRESHOLD) {
-            throw new IOException("Hadoop file system leak protection threshold has been reached. Threshold can be changed by setting " +
-                    "'-DhadoopFileSystemCacheLeakProtectionThreshold=<value>' jvm property. Current limit is: " + CACHE_LEAK_PROTECTION_THRESHOLD);
+        // Private credentials are not empty only in case of Kerberos authentication
```

Comment:
Add period at the end of the sentence.

This is unclear: "credentials are not empty only in case of Kerberos"

I think this is trying to say "credentials are only set when using Kerberos"